pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1542 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nftables
+
+//
+// NOTE: this needs to be tested in e2e since it uses iptables for everything.
+//
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base32"
+	"fmt"
+	"net"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/danwinship/nftables"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/events"
+	utilsysctl "k8s.io/component-helpers/node/util/sysctl"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/proxy"
+	"k8s.io/kubernetes/pkg/proxy/conntrack"
+	"k8s.io/kubernetes/pkg/proxy/healthcheck"
+	"k8s.io/kubernetes/pkg/proxy/metaproxier"
+	"k8s.io/kubernetes/pkg/proxy/metrics"
+	proxyutil "k8s.io/kubernetes/pkg/proxy/util"
+	proxyutiliptables "k8s.io/kubernetes/pkg/proxy/util/iptables"
+	"k8s.io/kubernetes/pkg/util/async"
+	utilexec "k8s.io/utils/exec"
+	netutils "k8s.io/utils/net"
+)
+
+const (
+	// the nftables table
+	kubeProxyTable = "kube-proxy"
+
+	// service dispatch
+	kubeServicesChain       = "services"
+	kubeServiceIPsMap       = "service-ips"
+	kubeServiceNodePortsMap = "service-nodeports"
+
+	// set of IPs that accept NodePort traffic
+	kubeNodePortIPsSet = "nodeport-ips"
+
+	// handling for services with no endpoints
+	kubeEndpointsCheckChain    = "endpoints-check"
+	kubeNoEndpointServicesMap  = "no-endpoint-services"
+	kubeNoEndpointNodePortsMap = "no-endpoint-nodeports"
+	kubeRejectChain            = "reject-chain"
+
+	// LoadBalancerSourceRanges handling
+	kubeFirewallSet             = "firewall"
+	kubeFirewallCheckChain      = "firewall-check"
+	kubeFirewallAllowSet        = "firewall-allow"
+	kubeFirewallAllowCheckChain = "firewall-allow-check"
+
+	// masquerading
+	kubeMarkMasqChain     = "mark-for-masquerade"
+	kubeMasqueradingChain = "masquerading"
+
+	// chain for special filtering rules
+	kubeForwardChain = "forward"
+)
+
+const sysctlNFConntrackTCPBeLiberal = "net/netfilter/nf_conntrack_tcp_be_liberal"
+
+// internal struct for string service information
+type servicePortInfo struct {
+	*proxy.BaseServicePortInfo
+	// The following fields are computed and stored for performance reasons.
+	nameString             string
```

Comment:
you may want to make them this variables typed as the iptables ones. `iptables.Chain` or `iptables.Table` to avoid subtle bugs with the names

Reply:
The table/chain types make sense for APIs like

```
EnsureRule(position RulePosition, table Table, chain Chain, args ...string)
```

where it's useful to syntactically distinguish the table name from the chain name, and both of them from the rule arguments.

But 99% of the time in `pkg/proxy/iptables/proxier.go`, we aren't using those APIs, and we end up doing:

```
				natRules.Write(
					"-A", string(externalTrafficChain),
					"-m", "comment", "--comment", fmt.Sprintf(`"masquerade traffic for %s external destinations"`, svcPortNameString),
					"-j", string(kubeMarkMasqChain))
```

ie, `iptables.Chain` just gets in our way and forces us to typecast it away. And it would be the same in the nftables proxy.

Reply:
good point